File: ruby/mastermind.rb
#!/usr/bin/ruby

level = 4
digits = 6
nbTurns = 12
turn = 1
secretCode = ""

for i in 1..level
    secretCode += (rand(digits) + 1).to_s
end

puts ""
puts "Welcome to Mastermind! (Ruby version)"
puts "The code is 4 digits long, chosen among 6 possible digits (1 to 6)"
puts "Can you find it in less than 12 turns?"
puts ""
# puts "DEBUG: " + secretCode
# puts ""

while turn < nbTurns do
    sCode = secretCode.dup
    hit = 0
    miss = 0

    puts "Turn #{ turn } - Your guess:"
    play = gets

    if play.length != 5 then
        puts "\nYou lose!\nAnswer was #{ secretCode }"
        exit
    end

    play.chomp.each_char.with_index do |c, i|
        if c == sCode[i] then
            hit += 1
            sCode[i] = 'X'
            play[i] = '#'
        end
    end

    play.chomp.each_char.with_index do |c, i|
        if sCode.include?(c) then
            miss += 1
            play[i] = '*'
        end
    end

    if hit == level then
        puts ""
        puts "You win! \\o/"
        puts
        exit
    end

    puts "# Right: #{ hit } - * Miss: #{ miss }"

    # Comment next block to disable hints
    play.chomp.each_char.with_index do |c, i|
        if c.match(/[[:digit:]]/) then
            play[i] = 'x'
        end
    end
    puts "Hint: #{ play } (#: Right, *: Miss, x: Wrong)"
    puts ""

    turn += 1
end

puts "\nYou lose!\nAnswer was #{ secretCode }"

File: js/mastermind.js
#!/usr/bin/node
// Simple Mastermind clone in js

const mastermind = function () {
  const readline = require('readline-sync');
  const level = 4;
  const digits = 6;
  const nbTurns = 12;
  let turn = 1;
  const secretCode = [];
  let code, right, miss;

  for (let i = 0; i < level; i++) {
    const rand = String(Math.floor(Math.random() * digits) + 1);
    secretCode.push(rand);
  }

  console.log('');
  console.log('Welcome to Mastermind! (Node.js version)');
  console.log('The code is 4 digits long, chosen among 6 possible digits (1 to 6)');
  console.log('Can you find it in less than 12 turns?');
  console.log('');
  // console.log('DEBUG: ' + secretCode.join(''));

  while (turn <= nbTurns) {
    code = secretCode.slice();
    right = 0;
    miss = 0;
    let play = readline.question(`Turn ${turn} - Your guess:\n`);
    play = Array.from(play);

    if (play.length !== level) {
      console.log(`\n(╯°□°)╯︵ ┻━┻\n\nYou lose!\nAnswer was ${secretCode.join('')}\n`);
      return;
    }

    for (let i = 0; i < level; i++) {
      if (code[i] === play[i]) {
        right++;
        code[i] = 'X';
        play[i] = '#';
      }
    }

    for (let i = 0; i < level; i++) {
      for (let j = 0; j < level; j++) {
        if (play[i] === code[j]) {
          miss++;
          play[i] = '*';
        }
      }
    }

    if (right === level) {
      console.log('\n(ง ͡ʘ ͜ʖ ͡ʘ)ง\n\nYou win! \\o/\n\n');
    }

    console.log(`# Right: ${right} - * Miss: ${miss}`);
    /* Comment next block to disable hints */
    for (let i = 0; i < level; i++) {
      if (play[i] !== '#' && play[i] !== '*') {
        play[i] = 'x';
      }
    }
    console.log(`Hint: ${play.join('')} (#: Right, *: Miss, x: Wrong)\n`);

    turn++;
  }
  console.log(`\n(╯°□°)╯︵ ┻━┻\n\nYou lose!\nAnswer was ${secretCode.join('')}\n`);
};

if (require.main === module) {
  mastermind();
}
